### PR TITLE
Optional directory listing

### PR DESCRIPTION
**Pull request by tetron**
Change handling of directory objects so "listing" isn't assumed.

This is an implementation of https://github.com/common-workflow-language/common-workflow-language/issues/348#issuecomment-266808096

### CODE REVIEW

**Comment by tetron** [comment]
Jenkins, test this please

File: cwltool/process.py [hunk 1]
```diff
@@ -44,7 +45,8 @@
                                 "ShellCommandRequirement",
                                 "StepInputExpressionRequirement",
                                 "ResourceRequirement",
-                                "InitialWorkDirRequirement"]
+                                "InitialWorkDirRequirement",
+                                "LoadListingRequirement"]
```
Comment (mr-c) [inline cwltool/process.py]: Can we have a test for using `LoadListingRequirement` without passing the `--enable-ext` flag? Should behave the same way `cwltool` does w/o this PR

File: cwltool/pathmapper.py [hunk 2]
```diff
@@ -106,6 +106,39 @@ def mark(d):
...
+                get_listing(fs_access, ent, recursive)
+            listing.append(ent)
+        else:
+            listing.append({"class": "File", "location": ld, "basename": bn})
+    rec["listing"] = listing
+
+def trim_listing(obj):
+    """Remove 'listing' field from Directory objects that are file references.
+
+    When Directory objects represent Keep references, it redundant and
```
Comment (mr-c) [inline cwltool/pathmapper.py]: Can you remove the reference to `Keep`?

File: cwltool/process.py (hunk 1 above)
Comment (tetron) [inline cwltool/process.py]: Added test_ext.TestListing which tests the various cases.

**Comment by tetron** [comment]
This is blocked on https://github.com/common-workflow-language/common-workflow-language/pull/400

**Pull request merged**